subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/AggregatingIncrementalAnnotationProcessingIntegrationTest.groovy

```diff
@@ -151,50 +160,146 @@ class AggregatingIncrementalAnnotationProcessingIntegrationTest extends Abstract
         def a = java "@Thing class A {}"
+        java "class Unrelated {}"
+
+        when:
         outputs.snapshot { succeeds "compileJava" }
 
+        then:
+        file("build/generated/sources/annotationProcessor/java/main/A.txt").exists()
+        file("build/headers/java/main/A.txt").exists()
+        file("build/classes/java/main/A.txt").exists()
+
+        when:
+        a.delete()
+        succeeds "compileJava"
+
+        then: "they all get cleaned"
+        outputs.deletedClasses("A")
+        !file("build/generated/sources/annotationProcessor/java/main/A.txt").exists()
+        !file("build/headers/java/main/A.txt").exists()
+        !file("build/classes/java/main/A.txt").exists()
+    }
+
+    def "SOURCE_OUTPUT resources get put into CLASS_OUTPUT if a location is not configured"() {
+        given:
+        buildFile << '''
+compileJava.options.setAnnotationProcessorGeneratedSourcesDirectory(null as File)
+'''
+        withProcessor(new ResourceGeneratingProcessorFixture().withOutputLocations(StandardLocation.SOURCE_OUTPUT.toString()).withDeclaredType(IncrementalAnnotationProcessorType.AGGREGATING))
+        java "@Thing class A {}"
+
         when:
-        a.text = "@Thing class A { void foo() {} }"
+        outputs.snapshot { succeeds "compileJava" }
 
         then:
-        succeeds "compileJava", "--info"
+        file("build/classes/java/main/A.txt").exists()
+    }
 
-        and:
-        outputContains("Full recompilation is required because an annotation processor generated a resource.")
+    def "resources will fail to be created if SOURCE_OUTPUT and CLASS_OUTPUT point to the same place on the file system"() {
+        given:
+        buildFile << '''
+compileJava.options.setAnnotationProcessorGeneratedSourcesDirectory(null as File)
+'''
+        withProcessor(new ResourceGeneratingProcessorFixture().withOutputLocations(StandardLocation.SOURCE_OUTPUT.toString(), StandardLocation.CLASS_OUTPUT.toString()).withDeclaredType(IncrementalAnnotationProcessorType.AGGREGATING))
+        java "@Thing class A {}"
+
+        when:
+        fails "compileJava"
+
+        then:
+        errorOutput.contains("Failed to generate resource file A.txt: Attempt to reopen a file for path")
+    }
+
+    def "NATIVE_HEADER_OUTPUT resources will fail with an exception if a location is not configured"() {
+        given:
+        buildFile << '''
+compileJava.options.headerOutputDirectory = null
+'''
+        withProcessor(new ResourceGeneratingProcessorFixture().withOutputLocations(StandardLocation.NATIVE_HEADER_OUTPUT.toString()).withDeclaredType(IncrementalAnnotationProcessorType.AGGREGATING))
+        java "@Thing class A {}"
+
+        when:
+        fails "compileJava"
+
+        then:
+        errorOutput.contains("Failed to generate resource file")
+    }
+
+    def "invalid locations are rejected by javac"() {
```

Comment:
This is testing javac, not Gradle, so doesn't need to be here.

Reply:
Will remove all of these you've commented on.  

In the case of this specific test, because we run the Location by javac first, we can't actually get integration test coverage of Gradle's validation.  Oh well.